In [ ]:

Q7. Sum of weights
Select all the "Innjoo" laptops from the dataset.
Select only columns RAM, Storage, Screen.
Get the underlying NumPy array. Let's call it X.
Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
Compute the inverse of XTX.
Create an array y with values [1100, 1300, 800, 900, 1000, 1100].
Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
What's the sum of all the elements of the result?
Note: You just implemented linear regression. We'll talk about it in the next lesson.

0.43
45.29
45.58
91.30


### Q1. Pandas version
What's the version of Pandas that you installed?

You can get the version information using the __version__ field:

`pd.__version__`

In [2]:
import pandas as pd

In [3]:
pd.__version__

'2.2.2'

### Q2. Records count
How many records are in the dataset?

* 12
* 1000
* __2160__
* 12160

In [4]:
!wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv

--2024-09-30 12:26:14--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 298573 (292K) [text/plain]
Saving to: ‘laptops.csv’

laptops.csv         100%[===================>] 291.58K  --.-KB/s    in 0.005s  

2024-09-30 12:26:15 (62.0 MB/s) - ‘laptops.csv’ saved [298573/298573]



In [6]:
df = pd.read_csv("laptops.csv")

In [8]:
df.shape # check number of records and columns

(2160, 12)

### Q3. Laptop brands
How many laptop brands are presented in the dataset?

* 12
* __27__
* 28
* 2160

In [9]:
df.columns

Index(['Laptop', 'Status', 'Brand', 'Model', 'CPU', 'RAM', 'Storage',
       'Storage type', 'GPU', 'Screen', 'Touch', 'Final Price'],
      dtype='object')

In [10]:
df['Brand'].value_counts()

Brand
Asus                415
HP                  368
Lenovo              366
MSI                 308
Acer                137
Apple               116
Dell                 84
Microsoft            77
Gigabyte             48
Razer                37
Medion               32
LG                   32
Alurin               29
PcCom                24
Samsung              22
Dynabook Toshiba     19
Vant                 11
Deep Gaming           8
Primux                8
Innjoo                6
Thomson               4
Prixton               3
Millenium             2
Denver                1
Jetwing               1
Realme                1
Toshiba               1
Name: count, dtype: int64

In [12]:
len(df['Brand'].unique())

27

### Q4. Missing values
How many columns in the dataset have missing values?

* 0
* 1
* 2
* __3__

In [13]:
df.isna().sum()

Laptop             0
Status             0
Brand              0
Model              0
CPU                0
RAM                0
Storage            0
Storage type      42
GPU             1371
Screen             4
Touch              0
Final Price        0
dtype: int64

### Q5. Maximum final price
What's the maximum final price of Dell notebooks in the dataset?

* 869
* 3691
* 3849
* __3936__

In [15]:
df[df['Brand'] == 'Dell']['Final Price'].max()

np.float64(3936.0)

### Q6. Median value of Screen
Find the median value of Screen column in the dataset.
Next, calculate the most frequent value of the same Screen column.
Use fillna method to fill the missing values in Screen column with the most frequent value from the previous step.
Now, calculate the median value of Screen once again.
Has it changed?

Hint: refer to existing mode and median functions to complete the task.

* Yes
* __No__

In [19]:
screen_mean = df['Screen'].median()
print(screen_mean)

15.6


In [23]:
screen_mode = df['Screen'].mode()
print(screen_mode)

0    15.6
Name: Screen, dtype: float64


In [33]:
mode = 15.6

In [34]:
filled_screen = df['Screen'].fillna(mode)

In [36]:
screen_mean_filled = filled_screen.median()
print(screen_mean_filled)

15.6


### Q7. Sum of weights
Select all the "Innjoo" laptops from the dataset.
Select only columns RAM, Storage, and Screen.
Get the underlying NumPy array. Let's call it X.
Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
Compute the inverse of XTX.
Create an array y with values [1100, 1300, 800, 900, 1000, 1100].
Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
What's the sum of all the elements of the result?
Note: You just implemented linear regression. We'll talk about it in the next lesson.

* 0.43
* 45.29
* 45.58
* __91.30__

In [37]:
df_train = df[df['Brand'] == 'Innjoo'][["RAM", "Storage", "Screen"]].reset_index(drop=True)

In [38]:
df_train

RAM  Storage  Screen
0    8      256    15.6
1    8      512    15.6
2    4       64    14.1
3    6       64    14.1
4    6      128    14.1
5    6      128    14.1

In [39]:
import numpy as np
X = np.array(df_train)

In [40]:
X

array([[  8. , 256. ,  15.6],
       [  8. , 512. ,  15.6],
       [  4. ,  64. ,  14.1],
       [  6. ,  64. ,  14.1],
       [  6. , 128. ,  14.1],
       [  6. , 128. ,  14.1]])

In [41]:
XTX = X.T @ X

In [42]:
XTX_inv = np.linalg.inv(XTX)

In [43]:
y = [1100, 1300, 800, 900, 1000, 1100]

In [44]:
w = XTX_inv @ X.T @ y

In [45]:
w

array([45.58076606,  0.42783519, 45.29127938])

In [46]:
print(w.sum())

91.2998806299555
